lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,832 @@
+  void CheckName(const parser::IfConstruct& If) {
+    const auto& Name{std::get<0>(std::get<0>(If.t).statement.t)};
+    if (!BothEqOrNone(Name, std::get<4>(If.t).statement.v)) {
+      EH.Report(Index, "IF construct name mismatch"_err_en_US);
+      NoErrors = false;
+    }
+    for (const auto& ElseIfBlock : std::get<2>(If.t)) {
+      const auto& E{std::get<0>(ElseIfBlock.t).statement.t};
+      if (!PresentAndEq(std::get<1>(E), Name)) {
+	EH.Report(Index, "ELSE IF statement name mismatch"_err_en_US);
+	NoErrors = false;
+      }
+    }
+    if (std::get<3>(If.t).has_value()) {
+      const auto& E{std::get<3>(If.t).value().t};
+      if (!PresentAndEq(std::get<0>(E).statement.v, Name)) {
+	EH.Report(Index, "ELSE statement name mismatch"_err_en_US);
+	NoErrors = false;
+      }
+    }
+  }
+  /// \brief Common code for SELECT CASE, SELECT RANK, and SELECT TYPE
+  template<typename A> void CheckName(const A& Case, const char *const Sel1,
+				      const char *const Sel2 = "") {
+    const auto& Name{std::get<0>(std::get<0>(Case.t).statement.t)};
+    if (!BothEqOrNone(Name, std::get<2>(Case.t).statement.v)) {
+      EH.Report(Index, "SELECT %s construct name mismatch"_err_en_US, Sel1);
+      NoErrors = false;
+    }
+    for (const auto& CS : std::get<1>(Case.t))
+      if (!PresentAndEq(std::get<1>(std::get<0>(CS.t).statement.t), Name)) {
+	EH.Report(Index, "%sCASE statement name mismatch"_err_en_US, Sel2);
+	NoErrors = false;
+      }
+  }
+
+  /// \brief Check <i>where-construct-name</i>
+  /// Constraint C1033 - opening and ending name must match if present, and
+  /// <i>masked-elsewhere-stmt</i> and <i>elsewhere-stmt</i> either match
+  /// or be unnamed
+  void CheckName(const parser::WhereConstruct& Where) {
+    const auto& Name{std::get<0>(std::get<0>(Where.t).statement.t)};
+    if (!BothEqOrNone(Name, std::get<4>(Where.t).statement.v)) {
+      EH.Report(Index, "WHERE construct name mismatch"_err_en_US);
+      NoErrors = false;
+    }
+    for (const auto& W : std::get<2>(Where.t))
+      if (!PresentAndEq(std::get<1>(std::get<0>(W.t).statement.t), Name)) {
+	EH.Report(Index,
+		  "ELSEWHERE (<mask>) statement name mismatch"_err_en_US);
+	NoErrors = false;
+      }
+    if (std::get<3>(Where.t).has_value()) {
+      const auto& E{std::get<3>(Where.t).value().t};
+      if (!PresentAndEq(std::get<0>(E).statement.v, Name)) {
+	EH.Report(Index, "ELSEWHERE statement name mismatch"_err_en_US);
+	NoErrors = false;
+      }
+    }
+  }
+
+  /// \brief Check constraint <i>construct-name</i> in scope (C1134 and C1166)
+  /// \param SStr  a string to specify the statement, \c CYCLE or \c EXIT
+  /// \param Label the name used by the \c CYCLE or \c EXIT
+  template<typename A> void CheckLabelContext(const char* const SStr,
+					      const A& Name) {
+    auto E{Names.crend()};
+    for (auto I{Names.crbegin()}; I != E; ++I) {
+      if (*I == Name)
+	return;
+    }
```

Comment:
Can you replace this loop with `std::find()`?

Reply:
OK, done. (Doesn't reduce lines of code, though.)